### PR TITLE
Add autostart script for mac

### PR DESCRIPTION
**Pull request by jbpratt**
Must be executed as root to write to LaunchAgents and runs as the $USER.
Starts `timeplot` at login

Didn't bother with spctl/Gatekeeper stuff to trust the application, this needs to be manually done.

I have tested locally and its working without issues.

### CODE REVIEW

**Comment by vn971** [comment]
Can we maybe adapt the plist file to be a "user agent"? https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html
> A user agent is essentially identical to a daemon, but is specific to a given logged-in user and executes only while that user is logged in.

This way, we probably don't need root, which can help trust and automation a lot.

I can integrate a correctly written plist into current timeplot codebase, but I cannot create and test the plist myself, untoftunately...

**Comment by jbpratt** [comment]
This is already using a ‘LaunchAgent’ which runs as the user currently
logged in. ‘LaunchDaemon’ runs as root. The script had to be executed as
root to write the Plist to ‘/Library/LaunchAgents/‘. This only happens once
on install.

edit: if you don't want the script at all, the plist can be pulled out and instructions be added to the README instead. Happy to change the PR.

**Comment by jbpratt** [comment]
Actually, I speak too soon. I am new to running this Mac so forgive me. It seems that `/Library/` stores all user configurations but each user also has a `$HOME/Library/`. Making the update now; no root permissions required :tada:

**Comment by vn971** [comment]
@jbpratt78 thanks for testing and researching it! That's good news.

One last part that can be done is to let `timeplot` itself insert the right `plist` file in the right directory, and call `launchctl load -w ...`.

This is how it works on Linux already, see https://github.com/vn971/timeplot/blob/e1d191bb6d2206e03f13aa9f62f123435ffcb557/src/main.rs#L577

All the rest is already there in the codebase, just need to implement the function `#[cfg(target_os = "macos")] fn add_to_autostart()` here: https://github.com/vn971/timeplot/blob/e1d191bb6d2206e03f13aa9f62f123435ffcb557/src/main.rs#L499

If done this way, the user can override the default config `create_autostart_entry=false` to `true`, and let timeplot itself create the "user agent" runner.

**Comment by vn971** [comment]
Would you be willing to work on this? Should be pretty-straightforward, similar to the Linux implementation nearby. What do you think?

**Comment by jbpratt** [comment]
@vn971 Just pushed the changes! Let me know if there are any issues with the code, I have tested it and it works just fine.

File: src/main.rs [hunk 1]
```diff
@@ -496,7 +496,33 @@ fn add_to_autostart() {
...
+
+	Command::new("launchctl")
+		.arg("load")
+		.arg("-w")
+		.arg(file_path)
+		.output()
+		.expect("failed to execute");
+}
+
+#[cfg(target_os = "windows")]
```
Comment (vn971) [inline src/main.rs]: Can you please also change this line to `#[cfg(all(not(target_os = "linux"), not(target_os = "macos")))]
`? I think it describes the case more precisely than pinning to windows. I have no clue whether anybody actually ran `timeplot` in e.g. OpenBSD, but maybe someone will?

File: src/main.rs [hunk 2]
```diff
@@ -496,7 +496,33 @@ fn add_to_autostart() {
...
+
+#[cfg(target_os = "macos")]
+fn add_to_autostart() {
+	let executable_name = env::current_exe().expect("failed to get current executable name");
+	let executable_name = executable_name.to_str().unwrap_or_else(|| {
+		panic!(
+			"failed to read executable name '{:?}' to string",
+			executable_name
+		)
+	});
```
Comment (vn971) [inline src/main.rs]: If the code is identical between the "macos" and other OS-es, maybe move it to a separate function, e.g.
```
fn executable_name()
```
?

**Review by vn971 — COMMENTED** [review]
Looks pretty nice! Two things would be nice to fix.

I'm not sure why travis CI is failing BTW, I don't see how it would be the PR-s fault. We can deal with that later, pretty sure there's no real code error here.

File: src/main.rs [hunk 3]
```diff
@@ -479,24 +479,49 @@ fn do_save_current(dirs: &ProjectDirs, image_dir: &PathBuf, conf: &Config) {
 	});
 }
 
+fn exe_name() -> Option<String> {
+	env::current_exe()
```
Comment (jbpratt) [inline src/main.rs:483]: Please let me know if there is a better way to do this

**Comment by jbpratt** [comment]
> Looks pretty nice! Two things would be nice to fix.
> 
> I'm not sure why travis CI is failing BTW, I don't see how it would be the PR-s fault. We can deal with that later, pretty sure there's no real code error here.

`rustup update` was failing locally as well as on travis I believe. But it seems to be working now.

**Comment by vn971** [comment]
@jbpratt78 Just merged. Thanks for the contribution!

> Please let me know if there is a better way to do this

One downside that I felt is that we swallow std::current_exe failure reason, which could be valuable if an error actually happens. On the other hand, we only later `unwrap` when we call this function, so I changed it to the more dirty approach of panic-ing right in place. We don't want to recover from failures of this function anyway. https://github.com/vn971/timeplot/commit/2f6d7a0d716aaeb2892f0d2ee818da51d06db097

I also added some very basic code to XML-escape the %PATH% inside plist file, and added you to Cargo.toml/authors list, since the change was non-trivial and pretty valuable (I can't test things on MacOS myself). I've used your most recent email alias, as shown in git history.

**Pull request merged**